File: apps/volunteer_hub/openapi_parser.py
from __future__ import annotations

from copy import deepcopy
from typing import Any


class OpenApiPostParser:
    """Parses an OpenAPI schema dictionary and extracts POST endpoint payload schemas."""

    def __init__(self, schema: dict[str, Any]):
        self.schema = schema

    def parse_post_endpoints(self) -> list[dict[str, Any]]:
        endpoints: list[dict[str, Any]] = []
        paths = self.schema.get("paths", {})

        for path, path_item in paths.items():
            post_operation = (path_item or {}).get("post")
            if not post_operation:
                continue

            request_body = post_operation.get("requestBody", {})
            content = request_body.get("content", {})
            media_type = self._pick_media_type(content)
            schema_node = (content.get(media_type) or {}).get("schema", {})

            endpoints.append(
                {
                    "path": path,
                    "operation_id": post_operation.get("operationId"),
                    "summary": post_operation.get("summary") or post_operation.get("description") or path,
                    "tags": post_operation.get("tags", []),
                    "media_type": media_type,
                    "schema": self._normalize_schema(schema_node),
                }
            )

        return endpoints

    def _pick_media_type(self, content: dict[str, Any]) -> str:
        if "application/json" in content:
            return "application/json"
        if content:
            return next(iter(content.keys()))
        return "application/json"

    def _normalize_schema(self, schema_node: dict[str, Any]) -> dict[str, Any]:
        if not schema_node:
            return {"type": "object", "properties": {}}

        resolved = self._resolve_refs(schema_node)
        if "allOf" in resolved:
            resolved = self._merge_all_of(resolved)

        # Keep parser output predictable for dynamic UI rendering.
        if resolved.get("type") != "object" and "properties" not in resolved:
            return {"type": "object", "properties": {"value": resolved}, "required": ["value"]}

        return resolved

    def _resolve_refs(self, value: Any, seen: set[str] | None = None) -> Any:
        if seen is None:
            seen = set()

        if isinstance(value, list):
            return [self._resolve_refs(item, seen) for item in value]

        if not isinstance(value, dict):
            return value

        if "$ref" in value:
            ref = value["$ref"]
            if ref in seen:
                return {}
            seen_with_ref = set(seen)
            seen_with_ref.add(ref)
            target = self._resolve_pointer(ref)
            return self._resolve_refs(target, seen_with_ref)

        return {key: self._resolve_refs(item, seen) for key, item in value.items()}

    def _resolve_pointer(self, ref: str) -> Any:
        if not ref.startswith("#/"):
            return {}

        pointer_parts = ref[2:].split("/")
        current: Any = self.schema

        for part in pointer_parts:
            token = part.replace("~1", "/").replace("~0", "~")
            if not isinstance(current, dict) or token not in current:
                return {}
            current = current[token]

        return deepcopy(current)

    def _merge_all_of(self, schema_node: dict[str, Any]) -> dict[str, Any]:
        merged: dict[str, Any] = {"type": "object", "properties": {}, "required": []}

        for item in schema_node.get("allOf", []):
            resolved_item = self._resolve_refs(item)
            if "allOf" in resolved_item:
                resolved_item = self._merge_all_of(resolved_item)

            for key, value in resolved_item.get("properties", {}).items():
                merged["properties"][key] = value

            for field in resolved_item.get("required", []):
                if field not in merged["required"]:
                    merged["required"].append(field)

        for key, value in schema_node.items():
            if key == "allOf":
                continue
            merged[key] = value

        return merged

File: apps/volunteer_hub/static/volunteer_hub/js/openapi_parser.js
function deepClone(value) {
    return value === undefined ? undefined : JSON.parse(JSON.stringify(value));
}

function decodePointerToken(token) {
    return token.replace(/~1/g, "/").replace(/~0/g, "~");
}

export function resolvePointer(schema, ref) {
    if (!ref || typeof ref !== "string" || !ref.startsWith("#/")) {
        return {};
    }

    const parts = ref.slice(2).split("/").map(decodePointerToken);
    let current = schema;

    for (const part of parts) {
        if (!current || typeof current !== "object" || !(part in current)) {
            return {};
        }
        current = current[part];
    }

    return deepClone(current);
}

function mergeRequired(base = [], extra = []) {
    return Array.from(new Set([...(base || []), ...(extra || [])]));
}

function mergeAllOf(schema, node, seenRefs) {
    const combined = {
        type: "object",
        properties: {},
        required: [],
    };

    for (const part of node.allOf || []) {
        const normalized = normalizeSchemaNode(schema, part, seenRefs);
        if (normalized.properties) {
            Object.assign(combined.properties, normalized.properties);
        }
        combined.required = mergeRequired(combined.required, normalized.required || []);
    }

    for (const [key, value] of Object.entries(node)) {
        if (key !== "allOf") {
            combined[key] = value;
        }
    }

    return combined;
}

export function normalizeSchemaNode(schema, node, seenRefs = new Set()) {
    if (!node || typeof node !== "object") {
        return { type: "object", properties: {} };
    }

    if (node.$ref) {
        if (seenRefs.has(node.$ref)) {
            return { type: "object", properties: {} };
        }
        const nextSeen = new Set(seenRefs);
        nextSeen.add(node.$ref);
        const resolved = resolvePointer(schema, node.$ref);
        return normalizeSchemaNode(schema, resolved, nextSeen);
    }

    if (node.allOf) {
        return normalizeSchemaNode(schema, mergeAllOf(schema, node, seenRefs), seenRefs);
    }

    if (node.oneOf && node.oneOf.length > 0) {
        return normalizeSchemaNode(schema, node.oneOf[0], seenRefs);
    }

    if (node.anyOf && node.anyOf.length > 0) {
        return normalizeSchemaNode(schema, node.anyOf[0], seenRefs);
    }

    const normalized = deepClone(node);

    if (normalized.properties && typeof normalized.properties === "object") {
        for (const [prop, child] of Object.entries(normalized.properties)) {
            normalized.properties[prop] = normalizeSchemaNode(schema, child, seenRefs);
        }
    }

    if (normalized.items) {
        normalized.items = normalizeSchemaNode(schema, normalized.items, seenRefs);
    }

    return normalized;
}

function pickRequestContent(content) {
    if (content["application/json"]) {
        return ["application/json", content["application/json"]];
    }

    const firstType = Object.keys(content || {})[0];
    if (firstType) {
        return [firstType, content[firstType]];
    }

    return ["application/json", { schema: { type: "object", properties: {} } }];
}

export function parseOpenApiPostEndpoints(schema) {
    const result = [];
    const paths = schema?.paths || {};

    for (const [path, pathItem] of Object.entries(paths)) {
        if (!pathItem?.post) {
            continue;
        }

        const operation = pathItem.post;
        const content = operation?.requestBody?.content || {};
        const [mediaType, mediaConfig] = pickRequestContent(content);
        const rawSchema = mediaConfig?.schema || { type: "object", properties: {} };

        result.push({
            path,
            method: "POST",
            summary: operation.summary || operation.description || path,
            operationId: operation.operationId || null,
            tags: operation.tags || [],
            mediaType,
            schema: normalizeSchemaNode(schema, rawSchema),
        });
    }

    return result;
}
